### ISSUE TITLE
Compatibility with Picard tools

### ISSUE THREAD

**Issue by Itolstoganov** [ksahlin/StrobeAlign#144]
Hi,

Currently, the default MAPQ value for the unmapped reads is 255, however, Picard tools require it to be 0. For instance, the command

`java -jar picard.jar AddOrReplaceReadGroups \ I=input.bam \ O=output.bam \ RGID=4 \ RGLB=lib1 \ RGPL=ILLUMINA \ RGPU=unit1 \ RGSM=20`

where input.bam was obtained by strobealign (with default parameters), throws an error message

`MAPQ should be 0 for unmapped read.`

The SAM specification does not set any particular MAPQ value for the unmapped reads, however, if the 255 value does not serve any specific purpose, built-in support for Picard tools might be beneficial.

**Comment by ksahlin** [comment]
Perhaps useful to change the MAPQ values of unmapped reads to 0 then. What does @marcelm think?

**Comment by marcelm** [comment]
TL; DR: Yes, we should change it for compatibility.

From what I can tell, StrobeAlign is correct and I don’t think Picard should complain. The SAM specification says this (emphasis mine):
> Bit 0x4 is the only reliable place to tell whether the read is unmapped. **If 0x4 is set, no assumptions
can be made about RNAME, POS, CIGAR, MAPQ, and bits 0x2, 0x100, and 0x800.**

To me, this means that Picard should ignore MAPQ for unmapped reads. And if it wants to enforce a value for MAPQ of an unmaped read, the only value that makes sense to me is 255 as "A value 255 indicates that the mapping quality is not available" (SAM spec).

That said, there’s no need for this value to be 255 in our case and we can just change it instead of taking it up with the Picard developers.

I got a few different errors when testing what Picard accepts. We should ensure we trigger none of them:

* MAPQ should be 0 for unmapped read
* MAPQ must be zero if RNAME is not specified
* Error parsing text SAM file. POS must be non-zero if RNAME is specified

I suggest we just run Picard during the tests to ensure we produce acceptable SAM/BAM output.

And of course we should also work on #31 so that using a separate tool for adding RG tags isn’t necessary anymore.

**Comment by marcelm** [comment]
Hi @Itolstoganov thanks for reporting! This is now changed in the main branch.

**Comment by Itolstoganov** [comment]
Hi @marcelm thanks!
